File: backend/routes/projects.py
from flask import Blueprint, request, jsonify, g
from db.database import get_db, rows_to_list
from auth.utils import require_auth, error_response
from auth.validation import ValidationErrors, require_str, optional_str, require_int, optional_int

bp = Blueprint('projects', __name__, url_prefix='/api/projects')


def project_with_meta(db, project_id):
    project = db.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
    if not project:
        return None

    owner = db.execute('SELECT id, name, email FROM users WHERE id = ?', (project['ownerId'],)).fetchone()
    members = db.execute(
        """SELECT u.id, u.name, u.email FROM project_members pm
           JOIN users u ON u.id = pm.userId WHERE pm.projectId = ?""",
        (project_id,),
    ).fetchall()
    counts_rows = db.execute(
        'SELECT status, COUNT(*) as count FROM tasks WHERE projectId = ? GROUP BY status', (project_id,)
    ).fetchall()

    counts = {'todo': 0, 'in-progress': 0, 'done': 0}
    for r in counts_rows:
        counts[r['status']] = r['count']
    total = sum(counts.values())

    result = dict(project)
    result['owner'] = dict(owner) if owner else None
    result['members'] = rows_to_list(members)
    result['taskCounts'] = counts
    result['totalTasks'] = total
    return result


def assert_access(db, project_id, user_id):
    """Returns (project_row, error_response)."""
    project = db.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
    if not project:
        return None, error_response(404, 'Project not found.')
    is_member = db.execute(
        'SELECT 1 FROM project_members WHERE projectId = ? AND userId = ?', (project_id, user_id)
    ).fetchone()
    if project['ownerId'] != user_id and not is_member:
        return None, error_response(403, "You don't have access to this project.")
    return project, None


@bp.get('')
@require_auth
def list_projects():
    db = get_db()
    rows = db.execute(
        """SELECT DISTINCT p.* FROM projects p
           LEFT JOIN project_members pm ON pm.projectId = p.id
           WHERE p.ownerId = ? OR pm.userId = ?
           ORDER BY p.createdAt DESC""",
        (g.user['id'], g.user['id']),
    ).fetchall()
    projects = [project_with_meta(db, p['id']) for p in rows]
    return jsonify({'projects': projects})


@bp.post('')
@require_auth
def create_project():
    data = request.get_json(silent=True) or {}
    try:
        title = require_str(data, 'title', 'Title', 3, 80)
        description = optional_str(data, 'description', 'Description', 500)
        workspace_id = optional_int(data, 'workspaceId', 'Workspace')
    except ValidationErrors as e:
        return jsonify({'errors': e.errors}), 422

    db = get_db()

    if workspace_id:
        role = db.execute(
            'SELECT workspaceRole FROM workspace_members WHERE workspaceId = ? AND userId = ?',
            (workspace_id, g.user['id']),
        ).fetchone()
        if not role:
            return error_response(403, "You don't have access to this workspace.")

    cur = db.execute(
        'INSERT INTO projects (title, description, ownerId, workspaceId) VALUES (?, ?, ?, ?)',
        (title, description, g.user['id'], workspace_id),
    )
    db.execute('INSERT INTO project_members (projectId, userId) VALUES (?, ?)', (cur.lastrowid, g.user['id']))
    db.commit()
    return jsonify({'project': project_with_meta(db, cur.lastrowid)}), 201


@bp.get('/<int:project_id>')
@require_auth
def get_project(project_id):
    db = get_db()
    project, err = assert_access(db, project_id, g.user['id'])
    if err:
        return err
    return jsonify({'project': project_with_meta(db, project_id)})


@bp.put('/<int:project_id>')
@require_auth
def update_project(project_id):
    db = get_db()
    project = db.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
    if not project:
        return error_response(404, 'Project not found.')
    if project['ownerId'] != g.user['id']:
        return error_response(403, 'Only the project owner can edit this project.')

    data = request.get_json(silent=True) or {}
    try:
        title = require_str(data, 'title', 'Title', 3, 80)
        description = optional_str(data, 'description', 'Description', 500)
    except ValidationErrors as e:
        return jsonify({'errors': e.errors}), 422

    db.execute('UPDATE projects SET title = ?, description = ? WHERE id = ?', (title, description, project_id))
    db.commit()
    return jsonify({'project': project_with_meta(db, project_id)})


@bp.delete('/<int:project_id>')
@require_auth
def delete_project(project_id):
    db = get_db()
    project = db.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
    if not project:
        return error_response(404, 'Project not found.')
    if project['ownerId'] != g.user['id']:
        return error_response(403, 'Only the project owner can delete this project.')
    db.execute('DELETE FROM projects WHERE id = ?', (project_id,))
    db.commit()
    return jsonify({'message': 'Project deleted.'})


@bp.post('/<int:project_id>/members')
@require_auth
def add_member(project_id):
    db = get_db()
    project = db.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
    if not project:
        return error_response(404, 'Project not found.')
    if project['ownerId'] != g.user['id']:
        return error_response(403, 'Only the project owner can add members.')

    data = request.get_json(silent=True) or {}
    try:
        user_id = require_int(data, 'userId', 'A valid user must be selected')
    except ValidationErrors as e:
        return jsonify({'errors': e.errors}), 422

    user = db.execute('SELECT id FROM users WHERE id = ?', (user_id,)).fetchone()
    if not user:
        return error_response(422, 'That user does not exist.', 'userId')

    try:
        db.execute('INSERT INTO project_members (projectId, userId) VALUES (?, ?)', (project_id, user_id))
        db.commit()
    except Exception:
        return error_response(422, 'That user is already a member.', 'userId')

    return jsonify({'project': project_with_meta(db, project_id)}), 201


@bp.delete('/<int:project_id>/members/<int:user_id>')
@require_auth
def remove_member(project_id, user_id):
    db = get_db()
    project = db.execute('SELECT * FROM projects WHERE id = ?', (project_id,)).fetchone()
    if not project:
        return error_response(404, 'Project not found.')
    if project['ownerId'] != g.user['id']:
        return error_response(403, 'Only the project owner can remove members.')
    if user_id == project['ownerId']:
        return error_response(422, 'The project owner cannot be removed.')
    db.execute('DELETE FROM project_members WHERE projectId = ? AND userId = ?', (project_id, user_id))
    db.commit()
    return jsonify({'project': project_with_meta(db, project_id)})

File: backend-node/routes/projects.js
const express = require('express');
const { body, param } = require('express-validator');
const db = require('../db/db');
const { handleValidation } = require('../middleware/validate');
const { requireAuth } = require('../middleware/auth');

const router = express.Router();
router.use(requireAuth);

function projectWithMeta(projectId, userId) {
  const project = db.prepare('SELECT * FROM projects WHERE id = ?').get(projectId);
  if (!project) return null;

  const owner = db.prepare('SELECT id, name, email FROM users WHERE id = ?').get(project.ownerId);
  const members = db
    .prepare(
      `SELECT u.id, u.name, u.email FROM project_members pm
       JOIN users u ON u.id = pm.userId WHERE pm.projectId = ?`
    )
    .all(projectId);
  const taskCounts = db
    .prepare(
      `SELECT status, COUNT(*) as count FROM tasks WHERE projectId = ? GROUP BY status`
    )
    .all(projectId);

  const counts = { todo: 0, 'in-progress': 0, done: 0 };
  taskCounts.forEach((r) => { counts[r.status] = r.count; });
  const total = counts.todo + counts['in-progress'] + counts.done;

  return { ...project, owner, members, taskCounts: counts, totalTasks: total };
}

function assertAccess(req, res, projectId) {
  const project = db.prepare('SELECT * FROM projects WHERE id = ?').get(projectId);
  if (!project) {
    res.status(404).json({ errors: [{ field: null, message: 'Project not found.' }] });
    return null;
  }
  const isMember = db
    .prepare('SELECT 1 FROM project_members WHERE projectId = ? AND userId = ?')
    .get(projectId, req.user.id);
  if (project.ownerId !== req.user.id && !isMember) {
    res.status(403).json({ errors: [{ field: null, message: "You don't have access to this project." }] });
    return null;
  }
  return project;
}

// List all projects the user owns or is a member of
router.get('/', (req, res) => {
  const rows = db
    .prepare(
      `SELECT DISTINCT p.* FROM projects p
       LEFT JOIN project_members pm ON pm.projectId = p.id
       WHERE p.ownerId = ? OR pm.userId = ?
       ORDER BY p.createdAt DESC`
    )
    .all(req.user.id, req.user.id);

  const projects = rows.map((p) => projectWithMeta(p.id, req.user.id));
  res.json({ projects });
});

const projectRules = [
  body('title')
    .trim()
    .notEmpty().withMessage('Title is required.')
    .isLength({ min: 3, max: 80 }).withMessage('Title must be between 3 and 80 characters.'),
  body('description')
    .optional({ checkFalsy: true })
    .isLength({ max: 500 }).withMessage('Description must be under 500 characters.'),
];

router.post('/', projectRules, handleValidation, (req, res) => {
  const { title, description = '' } = req.body;
  const info = db
    .prepare('INSERT INTO projects (title, description, ownerId) VALUES (?, ?, ?)')
    .run(title, description, req.user.id);
  db.prepare('INSERT INTO project_members (projectId, userId) VALUES (?, ?)').run(info.lastInsertRowid, req.user.id);

  res.status(201).json({ project: projectWithMeta(info.lastInsertRowid, req.user.id) });
});

router.get('/:id', param('id').isInt().withMessage('Invalid project id.'), handleValidation, (req, res) => {
  const project = assertAccess(req, res, req.params.id);
  if (!project) return;
  res.json({ project: projectWithMeta(project.id, req.user.id) });
});

router.put(
  '/:id',
  [param('id').isInt().withMessage('Invalid project id.'), ...projectRules],
  handleValidation,
  (req, res) => {
    const project = db.prepare('SELECT * FROM projects WHERE id = ?').get(req.params.id);
    if (!project) return res.status(404).json({ errors: [{ field: null, message: 'Project not found.' }] });
    if (project.ownerId !== req.user.id) {
      return res.status(403).json({ errors: [{ field: null, message: 'Only the project owner can edit this project.' }] });
    }

    const { title, description = '' } = req.body;
    db.prepare('UPDATE projects SET title = ?, description = ? WHERE id = ?').run(title, description, req.params.id);
    res.json({ project: projectWithMeta(req.params.id, req.user.id) });
  }
);

router.delete('/:id', param('id').isInt().withMessage('Invalid project id.'), handleValidation, (req, res) => {
  const project = db.prepare('SELECT * FROM projects WHERE id = ?').get(req.params.id);
  if (!project) return res.status(404).json({ errors: [{ field: null, message: 'Project not found.' }] });
  if (project.ownerId !== req.user.id) {
    return res.status(403).json({ errors: [{ field: null, message: 'Only the project owner can delete this project.' }] });
  }
  db.prepare('DELETE FROM projects WHERE id = ?').run(req.params.id);
  res.json({ message: 'Project deleted.' });
});

const memberRules = [
  param('id').isInt().withMessage('Invalid project id.'),
  body('userId').isInt().withMessage('A valid user must be selected.'),
];

router.post('/:id/members', memberRules, handleValidation, (req, res) => {
  const project = db.prepare('SELECT * FROM projects WHERE id = ?').get(req.params.id);
  if (!project) return res.status(404).json({ errors: [{ field: null, message: 'Project not found.' }] });
  if (project.ownerId !== req.user.id) {
    return res.status(403).json({ errors: [{ field: null, message: 'Only the project owner can add members.' }] });
  }
  const user = db.prepare('SELECT id FROM users WHERE id = ?').get(req.body.userId);
  if (!user) return res.status(422).json({ errors: [{ field: 'userId', message: 'That user does not exist.' }] });

  try {
    db.prepare('INSERT INTO project_members (projectId, userId) VALUES (?, ?)').run(req.params.id, req.body.userId);
  } catch {
    return res.status(422).json({ errors: [{ field: 'userId', message: 'That user is already a member.' }] });
  }
  res.status(201).json({ project: projectWithMeta(req.params.id, req.user.id) });
});

router.delete(
  '/:id/members/:userId',
  [param('id').isInt(), param('userId').isInt()],
  handleValidation,
  (req, res) => {
    const project = db.prepare('SELECT * FROM projects WHERE id = ?').get(req.params.id);
    if (!project) return res.status(404).json({ errors: [{ field: null, message: 'Project not found.' }] });
    if (project.ownerId !== req.user.id) {
      return res.status(403).json({ errors: [{ field: null, message: 'Only the project owner can remove members.' }] });
    }
    if (Number(req.params.userId) === project.ownerId) {
      return res.status(422).json({ errors: [{ field: null, message: 'The project owner cannot be removed.' }] });
    }
    db.prepare('DELETE FROM project_members WHERE projectId = ? AND userId = ?').run(req.params.id, req.params.userId);
    res.json({ project: projectWithMeta(req.params.id, req.user.id) });
  }
);

module.exports = router;
